File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.swift
import Foundation

func solution(_ dartResult:String) -> Int {
    var result: [Double] = []
    var buffer: String = ""
    
    for ch in dartResult {
        if Double(String(ch)) != nil { // 정수로 변환이 가능하다면
            buffer += String(ch)
            continue
        }
        
        if !buffer.isEmpty {
            result.append(Double(buffer)!)
            buffer = ""
        }
        
        switch ch {
        case "S":
            result[result.count - 1] = pow(result[result.count - 1], 1)
        case "D":
            result[result.count - 1] = pow(result[result.count - 1], 2)
        case "T":
            result[result.count - 1] = pow(result[result.count - 1], 3)
        case "*":
            if result.count <= 1 {
                result[result.count - 1] *= 2
            } else {
                result[result.count - 1] *= 2
                result[result.count - 2] *= 2
            }
        case "#":
            result[result.count - 1] = -result[result.count - 1]
        default:
            break
        }
    }
    
    return Int(result.reduce(0.0, +))
}

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>
#include <cmath>

using namespace std;

int solution(string dartResult) {
    int answer = 0, index = 0;
    string str = "";
    int arr[3] = {0, 0, 0};

    for (char ch : dartResult) {
        if ('0' <= ch && ch <= '9') {  // 숫자가 나올경우
            str += ch;
        } else if (ch == 'S') {
            arr[index] += stoi(str);
            str = "";
            index++;
        } else if (ch == 'D') {
            arr[index] += pow(stoi(str), 2);
            str = "";
            index++;
        } else if (ch == 'T') {
            arr[index] += pow(stoi(str), 3);
            str = "";
            index++;
        } else if (ch == '*') {
            if (index == 1) {
                arr[0] *= 2;
            } else {
                arr[index - 1] *= 2;
                arr[index - 2] *= 2;
            }
        } else if (ch == '#') {
            arr[index - 1] *= -1;
        }
    }

    answer += arr[0] + arr[1] + arr[2]; 
    return answer;
}